bftengine/tests/testViewChange/test_views_manager.cpp

```diff
@@ -0,0 +1,208 @@
+      rc.replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+  viewsManager->storeComplaint(std::move(complaint));
+
+  ASSERT_EQ(viewsManager->getAllMsgsFromComplainedReplicas().size(), 1);
+}
+
+TEST_F(ViewsManagerTest, form_a_quorum_of_complaints) {
+  for (int replicaId = 0; replicaId < rc.numReplicas; ++replicaId) {
+    if (replicaId == rc.replicaId) continue;
+
+    std::unique_ptr<ReplicaAsksToLeaveViewMsg> complaint(ReplicaAsksToLeaveViewMsg::create(
+        replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+    viewsManager->storeComplaint(std::move(complaint));
+  }
+
+  ASSERT_EQ(viewsManager->hasQuorumToLeaveView(), true);
+}
+
+TEST_F(ViewsManagerTest, status_message_with_complaints) {
+  bftEngine::impl::SeqNum lastExecutedSeqNum = 200;
+  constexpr bool viewIsActive = true;
+  constexpr bool hasNewChangeMsg = true;
+  constexpr bool listOfPPInActiveWindow = false, listOfMissingVCMsg = false, listOfMissingPPMsg = false;
+
+  for (int replicaId = 0; replicaId < rc.numReplicas; ++replicaId) {
+    if (replicaId == rc.replicaId) continue;
+
+    std::unique_ptr<ReplicaAsksToLeaveViewMsg> complaint(ReplicaAsksToLeaveViewMsg::create(
+        replicaId, initialView, ReplicaAsksToLeaveViewMsg::Reason::ClientRequestTimeout));
+
+    viewsManager->storeComplaint(std::move(complaint));
+  }
+
+  bftEngine::impl::ReplicaStatusMsg replicaStatusMessage(rc.getreplicaId(),
+                                                         initialView,
+                                                         lastStableSeqNum,
+                                                         lastExecutedSeqNum,
+                                                         viewIsActive,
+                                                         hasNewChangeMsg,
+                                                         listOfPPInActiveWindow,
+                                                         listOfMissingVCMsg,
+                                                         listOfMissingPPMsg);
+
+  viewsManager->addComplaintsToStatusMessage(replicaStatusMessage);
+
+  for (const auto& i : viewsManager->getAllMsgsFromComplainedReplicas()) {
+    ASSERT_TRUE(replicaStatusMessage.hasComplaintFromReplica(i.first));
+  }
+}
+
+TEST_F(ViewsManagerTest, get_quorum_for_next_view_on_view_change_message_with_enough_complaints) {
+  bftEngine::impl::ReplicaId sourceReplicaId = (rc.replicaId + 1) % rc.numReplicas;
+  std::set<bftEngine::impl::ReplicaId> otherReplicas;
+  constexpr int nextView = initialView + 1;
+  constexpr int viewToComplainAbout = initialView;
+
+  ViewChangeMsg viewChangeMsg = ViewChangeMsg(sourceReplicaId, nextView, lastStableSeqNum);
+
+  for (int i = 0; i < rc.numReplicas; ++i) {
+    if (i == rc.replicaId || i == sourceReplicaId) continue;
+    otherReplicas.insert(i);
+  }
+
+  ASSERT_LT(rc.fVal, otherReplicas.size());
```

Comment:
I think it would be more explicit and clearer to say `ASSERT_LE(rc.fVal+1, otherReplicas.size());`